docs: SAML attribute mapping

flyinghermit opened the pull request:
adds SAML attribute mapping reference docs. 

RFD: https://github.com/gravitational/teleport.e/blob/master/rfd/0018e-saml-idp-attribute-mapping.md
tracker: https://github.com/gravitational/teleport/issues/34725

flyinghermit committed:
saml attribute mapping reference

flyinghermit committed:
formatting

flyinghermit commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+
+
+## Testing attribute mapping
+### Prerequisites
```
@ptgott I added Prerequisites here as the requirements are only for testing commands below. But lmk if we prefer to keep it at top of the page.

r0mant replied:
I think prerequisites should go on top to be consistent with other guides.

flyinghermit committed:
rename test_attribute_mapping to test-attribute-mapping

flyinghermit commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+## Attribute value evaluation
+
+Attribute values are authored using Predicate expression. During SSO request, the expressions will be parsed from the
+Service Provider metadata and passed to Predicate evaluator for evaluation. The resulting values are asserted in the SAML
```
Good addition for clarity. Though I think the `authorized` should be `authenticated` as there isn't any authz happening in SSO request. Updated now as "Service Provider metadata and passed to Predicate evaluator for evaluation with the authorized user context."

flyinghermit commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+Split string at matching character.
+Expression | Result
+-----------| ------
+`strings.split(user.spec.traits.groups, "-", " ")` | `okta admin, dev sso, dev rdp`
```
Oh yes, thanks for catching that! I think I mixed up `regexp.replace` with `stings.split`. I was excluding `regexp.replace` (and in the process mixed it up with `stings.split`)  from reference as I was suggested we want to discourage using regexp.

flyinghermit commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+h1: SAML Attribute Mapping
+---
+
+Attribute mapping configures SAML Identity Provider to assert custom user attributes in SAML response.
```
I think `Attribute mapping configures Teleport SAML Identity Provider to assert custom user attributes in SAML response.` is a bit more clear. Notice I only added `Teleport` before `...SAML Identity Provider`.

flyinghermit committed:
- copy edits
- update add method to show set().add() and set(values) syntax
- update strings.split() syntax

flyinghermit commented:
Friendly ping.

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+h1: SAML Attribute Mapping
+---
+
+Attribute mapping configures Teleport SAML Identity Provider to assert custom user attributes in SAML response
```
```suggestion
Attribute mapping configures Teleport SAML Identity Provider to assert custom user attributes in SAML response.
```

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+Attribute mapping configures Teleport SAML Identity Provider to assert custom user attributes in SAML response
+The Teleport SAML IdP supports three configurable fields for attribute mapping:
+- `name`: Name of the outgoing attribute. Required. Name should be unique across attribute mapping.
+- `value`: Value defined using a predicate expression, which can reference Teleport usernames, roles or traits. Required.
```
Nit: Should we link out "predicate expression" to the [predicate language spec](https://goteleport.com/docs/reference/predicate-language/)?

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
````diff
@@ -0,0 +1,194 @@
+    - `basic`: value equals to `urn:oasis:names:tc:SAML:2.0:attrname-format:basic`.
+
+The values can be configured with `saml_idp_service_provider` spec.
+```
````
Nit: Specify format for proper syntax highlighting.
```suggestion
```yaml
```

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+    - `uri`: value equals to `urn:oasis:names:tc:SAML:2.0:attrname-format:uri`.
+    - `basic`: value equals to `urn:oasis:names:tc:SAML:2.0:attrname-format:basic`.
+
+The values can be configured with `saml_idp_service_provider` spec.
```
```suggestion
The values can be configured when adding a SAML application in Teleport web UI, or with `saml_idp_service_provider` resource spec created with `tctl create` or via API.
```

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+
+The following user attributes are available for mapping between Teleport IdP and Service Providers:
+
+| Attributes | Description                   | Syntax                                                       |
```
Nit: Feels like "description" column is really redundant in this table.

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+
+Predicate expressions for attribute mapping are evaluated against user attributes that can be accessed using evaluation context listed above.
+
+Expression syntax are similar to how expressions are used in [Login Rules](../login-rules/reference.mdx). But not all of the
```
Nit: I'm not sure we should mention login rules here tbh. This assumes readers have familiarity with login rules which may not be the case. I would just keep the sentence "The supported functions ...".

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
````diff
@@ -0,0 +1,194 @@
+Expression syntax are similar to how expressions are used in [Login Rules](../login-rules/reference.mdx). But not all of the
+functions and methods are available for attribute mapping. The supported functions and methods are listed below, along with
+the usage syntax and it's result, evaluated against the following reference user spec file:
+```
````
```suggestion
```yaml
```

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+    - editor
+    - dev-ssh
+  traits:
+    aws_role_arns: null
```
```suggestion
```

r0mant commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,194 @@
+
+- `--users`: user names or names of files containing user spec. Required.
+- `--sp`: name of file containing service provider spec with attribute mapping. Required.
+- `--format`: `yaml` or `json`.  Optional. Text output by default if the flag is not provided.
```
So it supports `text` output format too then?

flyinghermit replied:
It does if `--format` is not provided.
Now I think, I should have also added `text` in the `--format` : `--format: yaml | text | json` to avoid surprises.

r0mant reviewed (approved):
@flyinghermit I'm approving but please take a look at my comments before merging.

flyinghermit committed:
address review comments:
- move prerequisite up
- remove description from predicate evaluation context table
- copy edits

flyinghermit commented:
Friendly ping @ptgott @nklaassen @zmb3 @mdwn

ptgott reviewed (commented):
We should also add something to the section menu page, `docs/pages/access-controls/idps.mdx`.

ptgott commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,198 @@
+(!docs/pages/includes/no-oss-prereqs-tabs.mdx!)
+
+- (!docs/pages/includes/tctl.mdx!)
+- User with permission to create Service Provider resource. The preset `editor` role has this permission.
```
We introduce the Service Provider resource in one subsection of the SAML reference, which I'm not sure we can expect all readers of this page to have read. We should briefly explain this resource somewhere on this page.

flyinghermit replied:
Hmmm, so the feature "Attribute Mapping" is not a standalone feature but an extension of core SAML app configuration functionality. I mean user can surely bump straight into attribute mapping docs. But in practice, we can fairly expect users to have gone though our initial SAML guide, without which, attribute mapping itself is not possible.

ptgott commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,198 @@
+in Teleport.
+Attribute mapping that points to a non-existent value will not be included in SAML assertion.
+
+## Evaluation context
```
We should move this into the "Predicate expressions" section, otherwise it's difficult to tell what this is and how users would use it.

ptgott commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,198 @@
+
+(!docs/pages/includes/no-oss-prereqs-tabs.mdx!)
+
+- (!docs/pages/includes/tctl.mdx!)
```
```suggestion
```
I think we can remove this since this isn't a how-to guide and we don't expect users to be able to quickly copy and paste commands.

flyinghermit committed:
Merge branch 'master' into sshah/saml-attribute-mapping-docs

flyinghermit committed:
- add attribute mapping reference to idp.mdx (root page)
- add a line explaining the context of the page
- move attribute mapping context inside Predicate expressions heading
- remove tctl reference

flyinghermit commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
```diff
@@ -0,0 +1,198 @@
+description: How to map user attributes to custom SAML response
+h1: SAML Attribute Mapping
+---
+
```
https://github.com/gravitational/teleport/pull/36650/commits/ae18ffdf1d2c9077a282047bd59c3917b0d12bd5

flyinghermit committed:
revert previous commit on saml-attribute-mapping.mdx

mdwn commented on docs/pages/access-controls/idps/saml-attribute-mapping.mdx:
````diff
@@ -0,0 +1,205 @@
+      - dev-sso
+      - dev-rdp
+```
+### Methods
````
Given that we're using existing code for this bit, is there an existing doc page we could adapt for these methods?

flyinghermit replied:
The underlying functions and methods are borrowed from `traits_map` from login rules. But 
1. Not all the methods are exposed/works in the context attribute mapping. Adding new `Dict` values is not possible due to the fact that the final attribute value can be either a string or array of string, but not `Pair` value. So we have to hand pick method reference.
2. The reference example shown in login rules page is specific to login rules. But here, I wanted to show examples that are somewhat copy-pastable for attribute mapping. Without attribute mapping specific example, users will have to go through lots of guess work to adapt login rules syntax. 

I have also made reference to the underlying semantics minimal so we can point to a canonical predicate expression page for detailed syntax in future. But the example have to be in this page.

flyinghermit committed:
- move concept to top of the page
- add tctl reference back

flyinghermit committed:
single identation for sub nesting

flyinghermit committed:
single indent for attribute mapping name format

flyinghermit committed:
update saml reference copy on idp.mdx (root page)
